grpc/write_job.cpp

```diff
@@ -0,0 +1,164 @@
+	request.data_timestamp = to_dnet_time(fb_header->data_timestamp());
+	request.data_offset = fb_header->data_offset();
+	request.data_size = fb_header->data_size();
+	request.data_capacity = fb_header->data_capacity();
+	request.data_commit_size = fb_header->data_commit_size();
+	request.cache_lifetime = fb_header->cache_lifetime();
+
+	request.json = elliptics::data_pointer::allocate(request.json_size);
+	request.data = elliptics::data_pointer::allocate(request.data_size);
+}
+
+void deserialize_part(const write_job_t::rpc_request_t &rpc_request, bool first, write_job_t::request_t &request,
+	std::size_t &json_offset, std::size_t &data_offset) {
+
+	auto fb_request = rpc_request.GetRoot();
+	auto fb_header = fb_request->header();
+	if (first && fb_header) {
+		deserialize_header(fb_header, request);
+	}
+	put_data_part(request.json, json_offset, fb_request->json());
+	put_data_part(request.data, data_offset, fb_request->data());
+}
+
+write_job_t::rpc_response_t serialize(const write_job_t::response_t &response) {
+	flatbuffers::grpc::MessageBuilder builder;
+
+	auto json_timestamp = to_rpc_time(response.json_timestamp);
+	auto data_timestamp = to_rpc_time(response.data_timestamp);
+
+	auto fb_response = fb_grpc_dnet::CreateLookupResponse(
+		builder,
+		serialize_cmd(builder, response.cmd),
+		response.record_flags,
+		response.user_flags,
+		builder.CreateString(response.path),
+		&json_timestamp,
+		response.json_offset,
+		response.json_size,
+		response.json_capacity,
+		builder.CreateVector(response.json_checksum),
+		&data_timestamp,
+		response.data_offset,
+		response.data_size,
+		builder.CreateVector(response.data_checksum)
+	);
+
+	builder.Finish(fb_response);
+
+	return builder.ReleaseMessage<fb_grpc_dnet::LookupResponse>();
+}
+
+} // namespace
+
+class write_job_t::responder_t {
+public:
+	explicit responder_t(write_job_t &parent)
+	: parent_(parent) {
+	}
+
+	void reply(std::unique_ptr<response_t> response) {
+		parent_.send_response(std::move(response));
+	}
+
+private:
+	write_job_t &parent_;
+};
+
+write_job_t::write_job_t(dnet_node &node, ::grpc::ServerCompletionQueue &completion_queue,
+	fb_grpc_dnet::Elliptics::AsyncService &service)
+: async_reader_(&ctx_)
+, node_(node)
+, completion_queue_(completion_queue)
+, async_service_(service) {
+    proceed(true);
+}
+
+void write_job_t::proceed(bool more) {
+	// TODO: catch exceptions here
+	switch (state_) {
+	case state::CREATE:
+		state_ = state::REQUEST_RECEIVED_FIRST;
+		// TODO: Use constructor or clear memory for request
+		request_ = std::make_unique<request_t >();
+		async_service_.RequestWrite(&ctx_, &async_reader_, &completion_queue_, &completion_queue_, this);
+		break;
+	case state::REQUEST_RECEIVED_FIRST:
+		state_ = state::REQUEST_RECEIVED;
+		new write_job_t(node_, completion_queue_, async_service_);
+		read_next(true /*first*/, more);
+		break;
+	case state::REQUEST_RECEIVED:
+		read_next(false /*first*/, more);
+		break;
+	case state::FINISH:
+		delete this;
```

Comment:
Good manner is to put `break`/`return` at the ending case too. The same is for `read_job_t`.

Reply:
d22994a